### ISSUE TITLE
attribute-defined-outside-init false positive for attribute defined on subclass after verifying class type

### ISSUE THREAD

**Issue by Hawk777** [pylint-dev/pylint#9157]
### Bug description

```python
# pylint: disable=missing-docstring
# pylint: disable=too-few-public-methods

class Super:
    def __init__(self):
        pass


class Sub(Super):
    def __init__(self):
        super().__init__()
        self.some_value = 0


def get_super(x: bool) -> Super:
    if x:
        return Sub()
    return Super()


obj = get_super(True)
assert isinstance(obj, Sub)
obj.some_value = 27
```

### Command used

```shell
pylint test.py
```

### Pylint output

```python
************* Module test
test.py:23:0: W0201: Attribute 'some_value' defined outside __init__ (attribute-defined-outside-init)

------------------------------------------------------------------
Your code has been rated at 9.29/10 (previous run: 7.14/10, +2.14)
```

### Expected behavior

No errors: `some_value` is a member of `Sub`, defined in `Sub.__init__`, and we have by line 23 verified that `obj` is of type `Sub`. 

Note that this error is *not* triggered if `get_super` returns `Sub` unconditionally! It seems the `assert` is being ignored.

### Pylint version

```shell
pylint 3.0.1
astroid 3.0.1
Python 3.11.5 (main, Aug 29 2023, 19:39:22) [GCC 12.3.1 20230526]
```

**Comment by Hawk777** [comment]
This might be related to #8900, but that ticket is about union types while this is about superclasses and subclasses.
